### ISSUE TITLE
Can a FETCHed object be SUBSCRIBED to?

### ISSUE THREAD

**Issue by martinduke** [moq-wg/moq-transport#588]
This is a big one for relay and cache implementation.

If a relay receives an object via FETCH, is it possible to then serve it via SUBSCRIBE? This obviously partially depends on how SUBSCRIBE changes.

If the answer is yes, then caches need to keep information about subgroups (subgroup IDs for objects, last object in each subgroup for FIN purposes, etc). If not, they don't, and FETCH doesn't need to include that metadata.

**Comment by suhasHere** [comment]
cache store canonical objects and they have associated metadata .. isn't it same regardless of how the object arrives at the relay ?

Am i missing something ?

**Comment by wilaw** [comment]
With SUBSCRIBE only being future content, how would you be able to FETCH something, cache it and then SUBSCRIBE to it? I think the timing rules prevent this situation from occurring.

What is far more likely is that content initially SUBSCRIBED would later be FETCHed. This means you would not need to store subgroup metadata.

**Comment by englishm** [comment]
Yes, I'm inclined to think this is a simplification very much worth noting.

Once SUBSCRIBE is rewritten to only deal with forwarding newly arrived objects, I don't think there will remain circumstances where a relay would service a SUBSCRIBE using objects it retrieved with a FETCH.

(This assumes we can deal with the "join at current group" issue with something like a JOIN that a relay can locally decompose into an atomically aligned SUBSCRIBE and FETCH)

**Comment by afrind** [comment]
Individual Comment:

I agree that the current directions of FETCH and SUBSCRIBE indicate you can't subscribe to data in a cache.  However, if we add something like filters down the line, the relay may need to filter by subgroup.  I think we ought to keep the ids in FETCH wire format and caching model for now.

**Comment by englishm** [comment]
So long as we maintain that the data available in the SUBSCRIBE object wire format is a superset of what's needed to serve FETCHed objects, we preserve what's needed to achieve cache-ability and serve DVR or play behind live use cases from the objects SUBSCRIBE pulls through a relay.

Whether we actually trim fields from cached objects to store _only_ what's needed for FETCH is an optimization so I'm fine with @afrind's suggestion that we retain these fields in the FETCH wire format for now.

**Comment by martinduke** [comment]
If I SUBSCRIBE to something with a long-ish delivery timeout, should I get objects from cache that fall within that timeout?

If SUBSCRIBE can be for the current Group from zero, then FETCH can lead to SUBSCRIBE.

**Comment by martinduke** [comment]
Parking this because currently one can subscribe to the beginning of the current group. If that changes, we can revisit it.

**Comment by fluffy** [comment]
Yes, no matter how an object gets to a relay, the cacheing is the same.

**Comment by afrind** [comment]
Individual Comment: 

> Parking this because currently one can subscribe to the beginning of the current group. 

See #598

>  Yes, no matter how an object gets to a relay, the cacheing is the same.

+1

**Comment by suhasHere** [comment]
+1  on caching canonical object and regardless of how it arrives.

**Comment by englishm** [comment]
One reason we may want subgroup ids to be retained / transmitted with Fetch would be if we added a filtering feature to filter Fetch responses by subgroup id.

**Comment by fluffy** [comment]
> Parking this because currently one can subscribe to the beginning of the current group. If that changes, we can revisit it.

AFAIC that is not what the spec says

**Comment by afrind** [comment]
Chair Comment:

> AFAIC that is not what the spec says

This was true when the comment was written but not anymore - Current Group was removed from SUBSCRIBE in -09.

Individual Comment:

> One reason we may want subgroup ids to be retained / transmitted with Fetch would be if we added a filtering feature to filter Fetch responses by subgroup id.

I think there's a very valid use-case here - eg only FETCHing a particular layer - though we'd need to reconcile how subgroup filtered fetches work with gap detection.  And someone may figure how to allow SUBSCRIBE to VoD (Zafer's demo seemed to do this?) so maybe the answer to the origin issue is "yes, if the publisher knows how to pace the data".

In any case, I propose the resolution of this issue is that caches need to store all object metadata and serve it via FETCH.
